tty/cmsdk-apbuart/cmsdk-apbuart.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Phoenix-RTOS
+ *
+ * ARM CMSDK APB-UART driver
+ *
+ * Copyright 2024 Phoenix Systems
+ * Author: Lukasz Leczkowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <paths.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <board_config.h>
+#include <libtty.h>
+#include <libtty-lf-fifo.h>
+#include <libklog.h>
+
+#include <sys/debug.h>
+#include <sys/file.h>
+#include <sys/interrupt.h>
+#include <sys/mman.h>
+#include <sys/msg.h>
+#include <sys/platform.h>
+#include <sys/threads.h>
+
+#include <posix/utils.h>
+
+#include <phoenix/ioctl.h>
+
+
+#define UART_STACKSZ (1024)
+
+/* UART state bits */
+#define TX_BUF_FULL (1 << 0)
+#define RX_BUF_FULL (1 << 1)
+
+/* UART control bits */
+#define TX_EN     (1 << 0)
+#define RX_EN     (1 << 1)
+#define TX_INT_EN (1 << 2)
+#define RX_INT_EN (1 << 3)
+
+/* UART intstatus bits */
+#define TX_INT (1 << 0)
+#define RX_INT (1 << 1)
+
+#define UART_CLK SYSCLK_FREQ
+
+#define KMSG_CTRL_ID 100
+
+#ifndef UART_RXFIFOSIZE
+#define UART_RXFIFOSIZE 128
+#endif
+
+
+/* UART registers */
+/* clang-format off */
+enum { data = 0, state, ctrl, intstatus, bauddiv };
+/* clang-format on */
+
+
+typedef struct {
+	volatile uint32_t *base;
+	oid_t oid;
+
+	handle_t cond;
+	handle_t inth;
+	handle_t lock;
+	libtty_common_t tty;
+
+	/* fifo between irq and thread */
+	lf_fifo_t rxFifoCtx;
+	uint8_t rxFifoData[UART_RXFIFOSIZE];
+
+	uint8_t stack[UART_STACKSZ] __attribute__((aligned(8)));
+} uart_t;
+
+
+static const struct {
+	volatile uint32_t *base;
+	unsigned int rxirq;
+} info[] = {
+	{ .base = UART0_BASE, .rxirq = UART0_RX_IRQ },
+	{ .base = UART1_BASE, .rxirq = UART1_RX_IRQ },
+	{ .base = UART2_BASE, .rxirq = UART2_RX_IRQ },
+	{ .base = UART3_BASE, .rxirq = UART3_RX_IRQ },
+	{ .base = UART4_BASE, .rxirq = UART4_RX_IRQ },
+	{ .base = UART5_BASE, .rxirq = UART5_RX_IRQ },
+};
+
+
+static struct {
+	uart_t uart;
+	uint8_t stack[UART_STACKSZ] __attribute__((aligned(8)));
+} uart_common;
+
+
+static int uart_interrupt(unsigned int n, void *arg)
+{
+	uart_t *uart = (uart_t *)arg;
+	uint32_t status;
+	uint8_t c;
+
+	status = *(uart->base + state);
+
+	if ((status & RX_INT) != 0) {
+		while ((*(uart->base + state) & RX_BUF_FULL) != 0) {
+			c = *(uart->base + data) & 0xff;
+			lf_fifo_push(&uart->rxFifoCtx, c);
+		}
+		*(uart->base + intstatus) = RX_INT;
+
+		return 1;
+	}
+
+	return -1;
+}
+
+
+static void uart_intThread(void *arg)
+{
+	uart_t *uart = (uart_t *)arg;
+
+	for (;;) {
+		while (lf_fifo_empty(&uart->rxFifoCtx) != 0) { /* nothing to RX */
+			mutexLock(uart->lock);
```

Comment:
mutexLock/Unlock on each loop iteration - perhaps it would be better to optimize this a bit, perhaps if-while combo?